aria/storage/instrumentation.py

```diff
@@ -53,30 +59,76 @@ def track_changes(instrumented=None):
     :param instrumented: A dict from model columns to their python native type
     :return: The instrumentation context
     """
-    return _Instrumentation(instrumented or _INSTRUMENTED)
+    return _Instrumentation(ctx, instrumented or _INSTRUMENTED)
 
 
 class _Instrumentation(object):
 
-    def __init__(self, instrumented):
+    def __init__(self, ctx, instrumented):
         self.tracked_changes = {}
+        self.new_instances = {}
         self.listeners = []
+        self._new_modeled_instances = []
+        self._ctx = ctx
         self._track_changes(instrumented)
+        self._new_instance_index = 0
+
+    @property
+    def _new_instance_id(self):
+        self._new_instance_index += 1
+        return '{prefix}_{index}'.format(prefix=_NEW_INSTANCE, index=self._new_instance_index - 1)
+
+    def expunge_session(self):
```

Comment:
should also call expunge from rollback; also, what happens if the transaction is never completed?